server/src/main/java/org/elasticsearch/script/ScoreScriptUtils.java

```diff
@@ -53,27 +49,44 @@ public static double sigmoid(double value, double k, double a){
         return Math.pow(value,a) / (Math.pow(k,a) + Math.pow(value,a));
     }
 
-
     // reproducible random
-    public static double randomReproducible(String seedValue, int seed) {
-        int hash = StringHelper.murmurhash3_x86_32(new BytesRef(seedValue), seed);
+    public static double randomScore(String seedValue, String indexName, int shardId, int seed) {
+        int salt = (indexName.hashCode() << 10) | shardId;
+        int saltedSeed = mix32(salt ^ seed);
+        int hash = StringHelper.murmurhash3_x86_32(new BytesRef(seedValue), saltedSeed);
         return (hash & 0x00FFFFFF) / (float)(1 << 24); // only use the lower 24 bits to construct a float from 0.0-1.0
     }
 
-    // not reproducible random
-    public static final class RandomNotReproducible {
-        private final Random rnd;
-
-        public RandomNotReproducible() {
-           this.rnd = Randomness.get();
+    public static final class RandomScore {
+        private ScoreScript scoreScript;
+        public RandomScore(ScoreScript scoreScript) {
+            this.scoreScript = scoreScript;
         }
 
-        public double randomNotReproducible() {
-           return rnd.nextDouble();
+        public double randomScore(String seedValue, int seed) {
```

Comment:
Just for my clarification the seedValue here can change between documents?  Or rather is even expected to?  Is the seed also expected to change between documents or would it make more sense to consider the seed as read only for a script and have it be part of the constructor?

Also, now that this class binding has direct access to the ScoreScript reference, would it make more sense to have the user just pass in the field name for the seedValue and resolve it from the doc within this class to prevent the user from passing in an arbitrary String?

Reply:
@jdconrad  Right, `seedValue` is expected to be changed between documents, so different scores are produced for different documents.  And you are right about `seed` value, it would make more sense to have it as a part of the constructor. I will modify the code accordingly.

 

> would . it make more sense to have the user just pass in the field name for the seedValue and resolve it from the doc within this class to prevent the user from passing in an arbitrary String?

Good point! I will experiment with this.